### PR TITLE
fix(core): destroy detached `ViewRef`s when `ViewContainerRef` is destroyed

### PR DESCRIPTION
**Pull request by gabriel-quintiliano**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

The `ViewContainerRef` class had the following changes:
* Add `_detachedViews` : A private array property that keeps track of every `ViewRef` that was (and still is) detached from this `ViewContainerRef`.
* Add `_registerDestroyHookInHost`: Private method that will be responsible for adding a new destroy hook in host's TView as well as the proper context for calling this destroy hook in host's LView. In this case, destroy hook is `_destroyDetatchedViews` and the context is this `ViewContainerRef` instance.
* Add `_destroyDetatchedViews`: Private method that will be responsible for destroying all `ViewRefs` in `_detachedViews` array. This method will be called when this `ViewContainerRef`'s host component/directive is destroyed (and finally call its destroy hooks).
* Update `detach` method to also push the successfully detached view into `_detachedViews` array.
* Update `insert` method to also check if the view that will be inserted is among views of `_detachedViews` array and if so removes it from there (as it will no longer be detached).

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

When a `ViewContainerRef` is destroyed (because its host component/directive was destroyed) its currently detached views (`ViewRef`s previously detached from this `ViewContainerRef`) will NOT be destroyed which in some scenarios can cause memory leak issues as the OnDestroy hook of those detached views (and those in its view tree as well) won't run and thus subscriptions and other asynchronous operations wouldn't be cleared, as it's generally done within the `ngOnDestroy` hook method.

Issue Number: #51083

## What is the new behavior?

When a `ViewContainerRef` is destroyed (because its host component/directive was destroyed) its currently detached views (`ViewRef`s previously detached from this `ViewContainerRef`) will also get destroyed, unless they are attached to another container at that time.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/linker/view_container_ref.ts [hunk 1]
```diff
@@ -279,11 +279,41 @@ const VE_ViewContainerRef = ViewContainerRef;
...
+      hostLView[contextIdx as number] ??= context;
+    } else {
+      const emptySlot = hostLView.length;
+      hostLView[TVIEW].firstCreatePass && allocExpando(hostLView[TVIEW], hostLView, 1, null);
+      hostLView[emptySlot] = this;
+      destroyHooks.push(emptySlot, this._destroyDetatchedViews);
+    }
+  }
+
+  private _destroyDetatchedViews() {
```
Comment (ilirbeqirii) [inline packages/core/src/linker/view_container_ref.ts:312]: You seem to have a typo here: update "_destroyDetatchedViews" to "_destroyDetachedViews"

Comment (gabriel-quintiliano) [reply packages/core/src/linker/view_container_ref.ts:312]: oh, I'll take care of that, thanks a lot!

**Review by pkozlowski-opensource — CHANGES_REQUESTED** [review]
@gabriel-quintiliano I'm afraid that there are 2 problems with this PR:
* it doesn't have any test so it is not 100% clear what is the exact problem that we are fixing - especially that the linked issue is closed;
* I don't think that the logic / assumption is correct - detached views that were previously attached to the `ViewContainerRef` are... well... detached - in this sense `ViewContainerRef` don't "own" them any more and I don't think lifetime of those views should be linked to the `ViewContainerRef` - creator of those views should take care of destroying them.

**Comment by gabriel-quintiliano** [comment]
@pkozlowski-opensource I think you're right, I also didn't think it look 100% like it should be done. At first, I just wanted Angular to take care of the destroying those detached views, but now I see that the best is for the user to take care of that when dealing with dynamically created components.

Thank you for your review.

**Pull request closed without merging**